src/Cli/dotnet/commands/VirtualProjectBuildingCommand.cs

```diff
@@ -0,0 +1,192 @@
+            {
+                return 1;
+            }
+
+            // Then do a build.
+            var buildRequest = new BuildRequestData(
+                CreateProjectInstance(projectCollection),
+                targetsToBuild: ["Build"]);
+            var buildResult = BuildManager.DefaultBuildManager.BuildRequest(buildRequest);
+            if (buildResult.OverallResult != BuildResultCode.Success)
+            {
+                return 1;
+            }
+
+            BuildManager.DefaultBuildManager.EndBuild();
+            return 0;
+        }
+        catch (Exception e)
+        {
+            Console.Error.WriteLine(e.Message);
+            return 1;
+        }
+        finally
+        {
+            foreach (var (key, value) in savedEnvironmentVariables)
+            {
+                Environment.SetEnvironmentVariable(key, value);
+            }
+
+            binaryLogger?.Shutdown();
+            consoleLogger.Shutdown();
+        }
+
+        static ILogger? GetBinaryLogger(string[] args)
+        {
+            for (int i = args.Length - 1; i >= 0; i--)
+            {
+                var arg = args[i];
+                if (RunCommand.IsBinLogArgument(arg))
+                {
+                    return new BinaryLogger
+                    {
+                        Parameters = arg.IndexOf(':') is >= 0 and var index
+                            ? arg[(index + 1)..]
+                            : "msbuild.binlog",
+                    };
+                }
+            }
+
+            return null;
+        }
+    }
+
+    public ProjectInstance CreateProjectInstance(ProjectCollection projectCollection)
+    {
+        return CreateProjectInstance(projectCollection, addGlobalProperties: null);
+    }
+
+    private ProjectInstance CreateProjectInstance(
+        ProjectCollection projectCollection,
+        Action<IDictionary<string, string>>? addGlobalProperties = null)
+    {
+        var projectRoot = CreateProjectRootElement(projectCollection);
+
+        var globalProperties = projectCollection.GlobalProperties;
+        if (addGlobalProperties is not null)
+        {
+            globalProperties = new Dictionary<string, string>(projectCollection.GlobalProperties, StringComparer.OrdinalIgnoreCase);
+            addGlobalProperties(globalProperties);
+        }
+
+        return ProjectInstance.FromProjectRootElement(projectRoot, new ProjectOptions
+        {
+            GlobalProperties = globalProperties,
+        });
+    }
+
+    private ProjectRootElement CreateProjectRootElement(ProjectCollection projectCollection)
+    {
+        var projectFileFullPath = Path.ChangeExtension(EntryPointFileFullPath, ".csproj");
+        var projectFileText = """
+            <Project>
+                <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
+
+                <PropertyGroup>
+                    <OutputType>Exe</OutputType>
+                    <TargetFramework>net9.0</TargetFramework>
```

Comment:
Since this is in main this should be `net10.0` - we should have the current TFM available as a constant that you can interpolate into this structure.

Reply:
+1 to making this take into account current TF. Please don't resolve things you haven't fixed. I got here and planned to make pretty much exactly this comment.

Reply:
Sorry, I resolved this in a similar place in tests, didn't notice this comment was on a different place.